atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -37,41 +37,124 @@
+     * {@code rows} is irrelevant.
+     *
+     * If there are rows with no cells matching the provided {@link ColumnSelection}, they will not be present in the
+     * {@link Map#keySet()} of the output map at all. This accounts for local writes and deletes: a row written
+     * to locally (even if not persisted) will be present, and a row which is completely deleted locally (even if
+     * existing in the key value service) will be absent.
+     *
+     * @param tableRef table to load rows from
+     * @param rows rows to be loaded
+     * @param columnSelection columns to load from the given rows
+     * @return a mapping of rows to the columns matching the provided column selection
+     */
     @Idempotent
     SortedMap<byte[], RowResult<byte[]>> getRows(
             TableReference tableRef,
             Iterable<byte[]> rows,
             ColumnSelection columnSelection);
 
+    /**
+     * Returns a mapping of rows to {@link BatchingVisitable}s (which may be thought of as iterators over columns)
+     * within {@code tableRef} for the specified {@code rows}, where the columns fall within the provided
+     * {@link BatchColumnRangeSelection}. The single provided {@link BatchColumnRangeSelection} applies to all of the
+     * rows.
+     *
+     * The returned {@link BatchingVisitable}s are guaranteed to return cells matching the predicate. These are sorted
+     * by column, with ascending byte ordering.
+     *
+     * It is guaranteed that the {@link Map#keySet()} of the returned map has a corresponding element for each of the
+     * input {@code rows}, even if there are rows where no columns match the predicate.
+     *
+     * @param tableRef table to load values from
+     * @param rows unique rows to apply the column range selection to
+     * @param columnRangeSelection range of columns and batch size to load for each of the rows provided
+     * @return a mapping of rows to cells matching the predicate in the row, following the ordering outlined above
+     * @throws IllegalArgumentException if {@code rows} contains duplicates
+     */
     @Idempotent
     Map<byte[], BatchingVisitable<Map.Entry<Cell, byte[]>>> getRowsColumnRange(
             TableReference tableRef,
             Iterable<byte[]> rows,
             BatchColumnRangeSelection columnRangeSelection);
 
+    /**
+     * Returns a single iterator over the cell-value pairs in {@code tableRef} for the specified {@code rows}, where the
+     * columns fall within the provided {@link ColumnRangeSelection}. The single provided {@link ColumnRangeSelection}
+     * applies to all of the rows.
+     *
+     * If the provided {@link Iterable} of {@code rows} has a stable ordering, the returned iterator is guaranteed
+     * to return cell-value pairs in a lexicographic ordering over rows and columns, where rows are sorted according to
+     * the stable ordering of {@code rows}, and columns are sorted on byte ordering. If {@code rows} does not have a
```

Comment:
Why does iteration order of the passed iterable need to be stable? Can we not pick one format for the return order and stick to it?

Reply:
The return order for the stable case has to be the same order as was passed in (this was the core issue in PDS-115789, where an internal user did `getRowsColumnRange(table, [B, A], cols, batch)` and expected the cells for `B` to come back first). 

Generalising this to iterables without a consistent ordering seemed awkward. We _could_ guarantee something stronger (i.e. `if rows does not have a stable ordering, the returned iterator will return cell-value pairs where rows are sorted according to the order of the first iteration over it, and columns ...`), but it seems like an implementation detail I wouldn't want users to rely on, and it seems easy to break inadvertently.

Reply:
How about saying "Returned iterator is in the order of rows iterable then cells lexicographically. If your iterable does not have a stable iterator (i.e. iteration order can change between different calls to .iterator()) the cells are ordered lexicographically, but the order of rows is undefined."

As in I would put emphasis on the correct case, but explain that things will explode if that doesn't hold.

As a broader question: could we deprecate this method and automatically try to rewrite user's code (or rather suggest a rewritten version, to be reviewed by devs).

Reply:
First: Putting emphasis on the normal case makes sense.

Second question: I don't think so - this method is used in production in a number of places, and in some KVS implementations does something better than "concat the iterators from the Map returning methods"